### PR TITLE
Required changes to allow TorchScript export of PatchMatchNet module

### PR DESCRIPTION
**Pull request by anmatako**
This is the minimal set of changes needed to allow exporting of the PatchMatchNet module as a TorschScript binary. This export enables use of PatchMatchNet in colmap and potentially other relevant software. The change is backwards compatible, by avoiding the strict loading of the state dictionary. Main changes include:
- Flag to explicitly enable DataParallel in training since this is not supported in TorchScript by default
  - Module is now exported alongside the checkpoints when DataParallel is off
- Removal of some `del ...` commands inside branches since TorchScript incorrectly marks these variables as undefined
- Remove Tensor initialization to `None` since it's not supported by TorchScript. When null initialization is needed a `torch.empty(0)` is used instead and the "null" check is done by `numel() == 0`.
- Some function output changes to only use TorchScript supported types (e.g. no `Union`)
- Declaration of nn modules only happens inside `__init__`
- No conditional members in `__init__`. This may lead to unused modules being declared, but it has no adverse impact on funcitonality.

### CODE REVIEW

**Comment by anmatako** [comment]
I plan on subsequent changes for some cleanup and better compatibility with colmap. This should not be breaking any existing functionality.

**Review by FangjinhuaWang — COMMENTED** [review]
Thanks for your contribution! I roughly go through the changes and add some comments inside. Hope that you can have a test after commiting everything because I think I may miss something.

File: models/patchmatch.py [hunk 1]
```diff
@@ -183,9 +181,8 @@ def __init__(self, G: int = 8, stage: int = 3, evaluate_neighbors: int = 9, iter
 
         self.G = G
         self.stage = stage
-        if self.stage == 3:
-            self.pixel_wise_net = PixelwiseNet(self.G)
-
+        self.pixel_wise_net = PixelwiseNet(self.G)
```
[the remark this answers is not in the document]
Comment (anmatako) [reply models/patchmatch.py:184]: Yes, TorchScript cannot handle conditional members. It can be different between the branches, but it must exist for all branches. This effectively leads to a few unused members, but the functionality is not otherwise impacted.

File: models/patchmatch.py [hunk 2]
```diff
@@ -235,18 +232,18 @@ def forward(
...
-        if view_weights is not None:
+        if view_weights.numel() > 0:
             assert (
                 num_src_features == view_weights.size()[1]
             ), "Patchmatch Evaluation: Different number of images and view weights"
 
+        # Change to a tensor with value 1e-5
         pixel_wise_weight_sum = 1e-5
-
+        # pixel_wise_weight_sum = torch.zeros((batch, 1, 1, height, width), dtype=torch.float32, device=device)
```
[the remark this answers is not in the document]
Comment (anmatako) [reply models/patchmatch.py:242]: Yeah, this part is commented out and the initialization with 1e-5 actually works correctly. Sorry for missing this in the cleanup, I'll remove and update.

File: models/patchmatch.py [hunk 3]
```diff
@@ -387,20 +376,18 @@ def __init__(
...
-                    padding=self.dilation,
-                    dilation=self.dilation,
-                    bias=True,
-                )
-                nn.init.constant_(self.propa_conv.weight, 0.0)
-                nn.init.constant_(self.propa_conv.bias, 0.0)
+        # last iteration on stage 1 does not have propagation, but we still define this for TorchScript export compatibility
+        self.propa_conv = nn.Conv2d(
+            in_channels=self.propa_num_feature,
+            out_channels=max(2 * self.propagate_neighbors, 1),
```
[the remark this answers is not in the document]
Comment (anmatako) [reply models/patchmatch.py:382]: Yes, exactly. Yet another side-effect of TorchScript not handling conditional members. The extra one is unused anyway so it does not matter what is the actual number of channels.

File: models/module.py [hunk 4]
```diff
@@ -201,7 +201,7 @@ def depth_regression(p: torch.Tensor, depth_values: torch.Tensor) -> torch.Tenso
         result depth: expected value, soft argmin [B, 1, H, W]
     """
 
-    depth_values = depth_values.view(*depth_values.shape, 1, 1)
+    depth_values = depth_values.view(depth_values.shape[0], 1, 1)
```
[the remark this answers is not in the document]
Comment (FangjinhuaWang) [reply models/module.py:204]: The idea of regression is used in patchmatch.py. However, I do not call this certain function but only compute the expectation myself. You can have a check. I do not call this function actually.

[the remark this answers is not in the document]
Comment (anmatako) [reply models/module.py:204]: OK I think I see what you mean now. The only reference I can find for this function is during evaluation and only to calculate the photometric confidence, so it does not impact the model's training and the depth output in any way. Thanks!

**Comment by anmatako** [comment]
> Thanks for your contribution! I roughly go through the changes and add some comments inside. Hope that you can have a test after commiting everything because I think I may miss something.

I have tested with a smaller training and validation set and the results seem to be correct. Unfortunately, even with fixed seed in Torch I cannot get exactly identical results since the modules are slightly changed (removal of conditionals in member definitions), but I feel confident regarding the correctness of the change.

File: models/module.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (FangjinhuaWang) [reply models/module.py:204]: Ah, yes, you are correct. I forgot about "net.py". So your change is ok in this case since the depth values only have a shape of [D].

**Comment by FangjinhuaWang** [comment]
You mean there will be some different results in training? Actually, I forgot to fix the random seed for selecting source views during training (robust training strategy). However, even when I fix this random seed, the results may also slightly change when using different environments (such as different hardwares). So I think this is not an issue. Maybe you can also do evaluation on DTU using my trained model to see whether you can get same scores. If so, then I think it is fine.

**Comment by anmatako** [comment]
> You mean there will be some different results in training? Actually, I forgot to fix the random seed for selecting source views during training (robust training strategy). However, even when I fix this random seed, the results may also slightly change when using different environments (such as different hardwares). So I think this is not an issue. Maybe you can also do evaluation on DTU using my trained model to see whether you can get same scores. If so, then I think it is fine.

I will repeat the full DTU training and compare results using this vs the already existing checkpoint. I think if these match it should be enough confirmation that the changes are correct.

**Comment by anmatako** [comment]
@FangjinhuaWang I have compared the training from the updates I'm planning vs the training from the main branch and the results are matching. Note that I used Open3D for point cloud undersampling (0.4mm instead of 0.2mm in the paper) and nearest neighbor distance calculation to speed things up compared to using Matlab for these tasks.

Here are the overall outcomes from DTU evaluation:
- New with my changes: acc.: 0.483325, comp.: 0.395434, overall: 0.439380
- Current from main branch: acc.: 0.485234, comp.: 0.427619, overall: 0.456427

I think that at this point the change should be safe to merge.

**Comment by FangjinhuaWang** [comment]
Hi, Antonios. Great, then I will merge this. I may also do some test myself.
